### PR TITLE
EQL: Expand verification tests

### PR DESCRIPTION
**Pull request by aleksmaus**
Expand verification tests
Fix some error messaging consistency in EqlParser

Related to https://github.com/elastic/elasticsearch/issues/51873

@costin Could you take a look and let me know if this is what you were looking for that github issue.
Thanks!

### CODE REVIEW

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java [hunk 1]
```diff
@@ -58,8 +81,280 @@ public void testMisspelledColumnWithMultipleOptions() {
...
+        assertEquals("1:16: Unsupported function [arrayContains]",
+                errorParsing("registry where arrayContains(bytes_written_string_list, 'En')"));
+        assertEquals("1:16: Unsupported function [arraySearch]",
+                errorParsing("registry where arraySearch(bytes_written_string_list, a, a == 'en-us')"));
+        assertEquals("1:16: Unsupported function [arrayCount]",
+                errorParsing("registry where arrayCount(bytes_written_string_list, s, s == '*-us') == 1"));
+    }
+
+    // Some functions fail with "Unknown" message at the parse stage
+    // TODO (@costin): Should we make the error messaging consistent?
```
Comment (aleksmaus) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: @costin, I left this TODO comment, need your feedback on consistency for error messaging.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java [hunk 2]
```diff
@@ -58,8 +81,280 @@ public void testMisspelledColumnWithMultipleOptions() {
...
+        assertEquals("1:11: Unknown column [user_name_alia], did you mean any of [user_name, user_domain]?",
+                error(idxr, "foo where user_name_alia == 'bob'"));
+    }
+
+    // Test all elasticsearch numeric field types
+    public void testNumeric() {
+        final IndexResolution idxr = loadIndexResolution("mapping-numeric.json");
+        accept(idxr, "foo where long_field == 0");
+        accept(idxr, "foo where integer_field == " +
+                "0");
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: Single line, please.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java [hunk 3]
```diff
@@ -58,8 +81,280 @@ public void testMisspelledColumnWithMultipleOptions() {
...
+                error(idxr, "foo where ip_range_field == ''"));
+    }
+
+    public void testObject() {
+        final IndexResolution idxr = loadIndexResolution("mapping-object.json");
+        accept(idxr, "foo where endgame.pid == 0");
+
+        assertEquals("1:11: Unknown column [endgame.pi], did you mean [endgame.pid]?",
+                error(idxr, "foo where endgame.pi == 0"));
+
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: Please, remove extra empty line.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java [hunk 4]
```diff
@@ -58,8 +81,280 @@ public void testMisspelledColumnWithMultipleOptions() {
...
+        final IndexResolution idxr = loadIndexResolution("mapping-join.json");
+        accept(idxr, "foo where serial_event_id == 0");
+    }
+
+    public void testMultiField() {
+        final IndexResolution idxr = loadIndexResolution("mapping-multi-field.json");
+        accept(idxr, "foo where multi_field.raw == 'bar'");
+
+        assertEquals("1:11: [multi_field.english == 'bar'] cannot operate on first argument field of data type [text]: " +
+                        "No keyword/multi-field defined exact matches for [english]; define one or use MATCH/QUERY instead",
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: I think this error message is wrong, unless EQL plans to support MATCH/QUERY predicates.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java [hunk 5]
```diff
@@ -58,8 +81,280 @@ public void testMisspelledColumnWithMultipleOptions() {
...
+        accept(idxr, "foo where multi_field_options.key == 'bar'");
+
+        accept(idxr, "foo where multi_field_ambiguous.one == 'bar'");
+        accept(idxr, "foo where multi_field_ambiguous.two == 'bar'");
+        assertEquals("1:11: [multi_field_ambiguous.normalized == 'bar'] cannot operate on first argument field of data type [keyword]: " +
+                        "Normalized keyword field cannot be used for exact match operations",
+                error(idxr, "foo where multi_field_ambiguous.normalized == 'bar'"));
+
+        assertEquals("1:11: [multi_field_nested.dep_name == 'bar'] cannot operate on first argument field of data type [text]: " +
+                        "No keyword/multi-field defined exact matches for [dep_name]; define one or use MATCH/QUERY instead",
```
Comment (astefan) [inline x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: Same here about MATCH/QUERY predicates.

**Review by astefan — CHANGES_REQUESTED** [review]
Nice addition of tests.
I left two comments about MATCH/QUERY predicates inside an error message.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java (hunk 3 above)
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: I kind of liked the one line visual separation between successful and failure cases in here

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java (hunk 4 above)
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: @costin

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java (hunk 3 above)
Comment (astefan) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: This empty line is at the end of the method, before the closing `}` bracket...

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: @aleksmaus please raise an issue and mark it as discuss to get consensus about how we deal with exact matches on analyzed fields - whether we prevent any exact comparison against them or introduce some other type of comparison (inexact matches - say `~=`).

**Review by costin — APPROVED** [review]
Looks really good. Thanks Aleks!

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java (hunk 5 above)
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: Discussed with Costin, the behavior is correct, the error message can be improved in the context of EQL. Agreed  on leaving the test as is and open an "issue" for error messaging improvement for this particular case. Possibly can bring up for discussion for our weekly.

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java (hunk 2 above)
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: will update

File: x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java (hunk 3 above)
Comment (aleksmaus) [reply x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/VerifierTests.java]: will fix

**Review by astefan — APPROVED** [review]
LGTM

**Pull request merged**